---
title: "Filter Mutation Frequencies Tables"
output: 
  html_notebook:
    toc: TRUE
    toc_float: TRUE
    toc_depth: 4
author: Eric Wafula for Pediatric OpenTargets
date: 01/10/2021
---

Purpose: Remove Ensembl (ESNG) gene identifier in the mutation frequency tables, including SNV, CNV and fusion that are not in GENCODE v39 and Ensembl package 104.


## Usage

To run this from the command line, use:
```
Rscript -e "rmarkdown::render('filter-mutation-frequencies-tables.Rmd', clean = TRUE)"
```
_This assumes you are in the modules directory of the repository, OpenPedCan-analysis/analyses/filter-mutation-frequency-tables._

## Setup

Load libraries
```{r}
# R analysis packages
suppressWarnings(
  suppressPackageStartupMessages(library(rtracklayer))
)
suppressPackageStartupMessages(library(tidyverse))
suppressPackageStartupMessages(library(data.table))

# Magrittr pipe
`%>%` <- dplyr::`%>%`
```


Set up directories. 
```{r}
# directories for input and output files
root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))
data_dir <- file.path(root_dir, "data")
analyses_dir <- file.path(root_dir, "analyses")
module_dir <- file.path(analyses_dir, "filter-mutation-frequencies-tables")
gene_match_input_dir <- file.path(analyses_dir,  "gene_match/input")
snv_results_dir <- file.path(analyses_dir,  "snv-frequencies/results")
cnv_results_dir <- file.path(analyses_dir,  "cnv-frequencies/results")
fusion_results_dir <- file.path(analyses_dir,  "fusion-frequencies/results")
input_dir <- file.path(module_dir, "input")
results_dir <- file.path(module_dir, "results")

# Create results folder if it doesn't exist
if (!dir.exists(results_dir)) {
  dir.create(results_dir)
}
```

## Get input data

Load GENECODE gtf file
```{r}
# read file
gtf <- file.path(data_dir, "gencode.v39.primary_assembly.annotation.gtf.gz")
gencode_gtf <- rtracklayer::import(con = gtf) %>% as.data.frame() %>% tibble::as_tibble()
```

Get GENCODE Ensembl identifiers
```{r}
# filter for gene rows and get ids
ensg_ids <- gencode_gtf %>% 
  dplyr::filter(type == "gene") %>% 
  dplyr::mutate(gene_id = stringr::str_extract(gene_id, "ENSG\\d+")) %>% 
  pull(gene_id) %>% 
  unique() 
ensg_ids <- ensg_ids[!is.na(ensg_ids)]

# Remove gencode dataframe
rm(gencode_gtf)
```

Get deprecated Open Targets Ensembl identifiers
```{r}
# load OT deprecated ensembl IDs 
ot_ensg_ids <- readr::read_lines(file.path(input_dir, "MTP_v11_InvalidENSG_20220831.txt"))
```

Get PMTL Ensembl identifiers
```{r}
# filter for gene rows and get ids
pmtl_df <- readr::read_tsv(file.path(gene_match_input_dir, "PMTL_v1.1.tsv"))
pmtl_ids <-pmtl_df %>% 
  pull(Ensembl_ID) %>% 
  unique() 
# Remove PMTL dataframe
rm(pmtl_df)
```

Get OpenPedCan sample IDs (Kids_First_Biospecimen_ID)
```{r}
sample_ids <- readr::read_tsv(file.path(data_dir, "histologies.tsv"), 
                              guess_max = 10000) %>% 
  dplyr::filter(sample_type == "Tumor") %>% 
  dplyr::select(Kids_First_Biospecimen_ID, cohort)
```

## Filter mutation frequencies tables

Function to filter for required Ensembl gene identifiers and write to file
```{r}

filter_freq_table <- function(freq_file, ensg_ids, ot_ensg_ids) {
  
  # Read gene level table 
  freq_table <-  readr::read_tsv(freq_file, guess_max = 10000) 
  
  # Write filtered TSV file
  file_name <- basename(freq_file)
  freq_table %>% dplyr::filter(targetFromSourceId %in% ensg_ids) %>%
    dplyr::filter(!targetFromSourceId %in% ot_ensg_ids) %>%
    readr::write_tsv(file.path(results_dir, file_name))

  # Write filtered JSON file
  json_file <- paste(unlist(str_split(file_name, "\\."))[1], "json", sep = ".")
  freq_table %>% dplyr::filter(targetFromSourceId %in% ensg_ids) %>%
    dplyr::filter(!targetFromSourceId %in% ot_ensg_ids) %>%
    jsonlite::write_json(file.path(results_dir,  json_file))  

return(freq_table)
}
```

#### Gene level SNV frequencies
```{r}
# Remove Ensembl gene identifiers
gene_level_snv <- filter_freq_table(
  file.path(snv_results_dir, 
            "gene-level-snv-consensus-annotated-mut-freq.tsv.gz"), ensg_ids, ot_ensg_ids) 

# Ensembl gene identifiers removed
filtered_ensg <- gene_level_snv %>% 
  dplyr::filter(!targetFromSourceId %in% setdiff(ensg_ids, ot_ensg_ids)) %>% 
  dplyr::select(Gene_symbol, targetFromSourceId) %>%
  dplyr::rename(Ensembl_ID = targetFromSourceId) %>% 
  dplyr::distinct()

# read the SNV consensus file
consensus_ensg <- data.table::fread(
  file.path(data_dir, "snv-consensus-plus-hotspots.maf.tsv.gz"), 
  select = c("Tumor_Sample_Barcode", "Gene"), showProgress = FALSE) %>% 
  tibble::as_tibble() %>% 
  dplyr::filter(!is.na(Gene)) %>% 
  dplyr::rename(Kids_First_Biospecimen_ID = Tumor_Sample_Barcode) %>% 
  dplyr::distinct()

# merge sample IDs to the ENSG IDs not in gencode v39
filtered_ensg <- filtered_ensg %>% 
  dplyr::left_join(consensus_ensg, by=c("Ensembl_ID" = "Gene"))

# merge OpenPedCan cohorts to the ENSG IDs not in gencode v39
filtered_ensg <- filtered_ensg %>% 
  dplyr::left_join(sample_ids, by= "Kids_First_Biospecimen_ID")

# write to file
readr::write_tsv(filtered_ensg,
  file.path(results_dir, 
            "gene-level-snv-consensus-annotated-mut-freq_dropped_ensg.tsv.gz"))

# Display removed ENSEMBL IDs 
filtered_ensg %>% dplyr::select(Ensembl_ID, cohort) %>% 
  dplyr::distinct() %>%
  tidyr::drop_na() %>%  
  table()

# Display samples in removed ENSEMBL IDs that are in PMTL list
filtered_ensg %>% dplyr::filter(Ensembl_ID %in% pmtl_ids) %>% 
  select(Ensembl_ID, Kids_First_Biospecimen_ID) %>% 
  dplyr::distinct()

# Remove data frames
rm(gene_level_snv, filtered_ensg)
```

#### Variant level SNV frequencies
```{r}
# Remove Ensembl gene identifiers
variant_level_snv <- filter_freq_table(
  file.path(snv_results_dir, 
            "variant-level-snv-consensus-annotated-mut-freq.tsv.gz"), ensg_ids, ot_ensg_ids) 

# Ensembl gene identifiers removed
filtered_ensg <- variant_level_snv %>% 
  dplyr::filter(!targetFromSourceId %in% setdiff(ensg_ids, ot_ensg_ids)) %>% 
  dplyr::select(Gene_symbol, targetFromSourceId) %>%
  dplyr::rename(Ensembl_ID = targetFromSourceId) %>% 
  dplyr::distinct()

# merge sample IDs to the ENSG IDs not in gencode v39
filtered_ensg <- filtered_ensg %>% 
  dplyr::left_join(consensus_ensg, by=c("Ensembl_ID" = "Gene"))

# merge OpenPedCan cohorts to the ENSG IDs not in gencode v39
filtered_ensg <- filtered_ensg %>% 
  dplyr::left_join(sample_ids, by= "Kids_First_Biospecimen_ID")

# write to file
readr::write_tsv(filtered_ensg,
  file.path(results_dir, 
            "variant-level-snv-consensus-annotated-mut-freq_dropped_ensg.tsv.gz"))

# Display removed ENSEMBL IDs 
filtered_ensg %>% dplyr::select(Ensembl_ID, cohort) %>% 
  dplyr::distinct() %>%
  tidyr::drop_na() %>% 
  table()

# Display samples in removed ENSEMBL IDs that are in PMTL list
filtered_ensg %>% dplyr::filter(Ensembl_ID %in% pmtl_ids) %>% 
  select(Ensembl_ID, Kids_First_Biospecimen_ID) %>% 
  dplyr::distinct()

# Remove data frames
rm(variant_level_snv, filtered_ensg, consensus_ensg)
```

#### Gene level CNV frequencies
```{r}
# Remove Ensembl gene identifiers
gene_level_cnv <- filter_freq_table(
  file.path(cnv_results_dir, 
            "gene-level-cnv-consensus-annotated-mut-freq.tsv.gz"), ensg_ids, ot_ensg_ids) 

# Ensembl gene identifiers removed
filtered_ensg <- gene_level_cnv %>% 
  dplyr::filter(!targetFromSourceId %in% setdiff(ensg_ids, ot_ensg_ids)) %>% 
  dplyr::select(Gene_symbol, targetFromSourceId) %>%
  dplyr::rename(Ensembl_ID = targetFromSourceId) %>% 
  dplyr::distinct()

# read the CNV consensus file
consensus_ensg <- data.table::fread(
  file.path(data_dir, "consensus_wgs_plus_cnvkit_wxs.tsv.gz"), 
  select = c("biospecimen_id", "ensembl"), showProgress = FALSE) %>% 
  tibble::as_tibble() %>% 
  dplyr::filter(!is.na(ensembl)) %>% 
  dplyr::rename(Kids_First_Biospecimen_ID = biospecimen_id) %>% 
  dplyr::distinct()

# merge sample IDs to the ENSG IDs not in gencode v39
filtered_ensg <- filtered_ensg %>% 
  dplyr::left_join(consensus_ensg, by=c("Ensembl_ID" = "ensembl"))

# merge OpenPedCan cohorts to the ENSG IDs not in gencode v39
filtered_ensg <- filtered_ensg %>% 
  dplyr::left_join(sample_ids, by= "Kids_First_Biospecimen_ID")

# write to file
readr::write_tsv(filtered_ensg,
  file.path(results_dir, 
            "gene-level-cnv-consensus-annotated-mut-freq_dropped_ensg.tsv.gz"))

# Display removed ENSEMBL IDs 
filtered_ensg %>% dplyr::select(Ensembl_ID, cohort) %>% 
  dplyr::distinct() %>%
  tidyr::drop_na() %>% 
  table()

# Display samples in removed ENSEMBL IDs that are in PMTL list
filtered_ensg %>% dplyr::filter(Ensembl_ID %in% pmtl_ids) %>% 
  select(Ensembl_ID, Kids_First_Biospecimen_ID) %>% 
  dplyr::distinct()

# Remove data frames
rm(gene_level_cnv, filtered_ensg, consensus_ensg)
```

#### Fusion frequencies
```{r}
# Remove Ensembl gene identifiers
fusion <- filter_freq_table(
  file.path(fusion_results_dir, 
            "putative-oncogene-fusion-freq.tsv.gz"), ensg_ids, ot_ensg_ids) 

# Ensembl gene identifiers removed
filtered_ensg <- fusion %>% 
  dplyr::filter(!targetFromSourceId %in% setdiff(ensg_ids, ot_ensg_ids)) %>% 
  dplyr::select(FusionName, Gene_symbol, targetFromSourceId) %>%
  dplyr::rename(Ensembl_ID = targetFromSourceId) %>% 
  dplyr::distinct()

# split fusion name into constituent genes  symbols
fusion_names <- data.table::fread(
  file.path(data_dir, "fusion-putative-oncogenic.tsv"), 
  select = c("Sample", "FusionName"), showProgress = FALSE) %>% 
  tibble::as_tibble() %>% 
  tidyr::separate_rows(FusionName, sep = "--", convert = FALSE) %>% 
  tidyr::separate_rows(FusionName, sep = '/', convert = FALSE) %>% 
  dplyr::rename(Kids_First_Biospecimen_ID = Sample, 
                Gene_symbol = FusionName) %>% 
  dplyr::filter(!is.na(Gene_symbol)) %>% 
  dplyr::distinct() 

# merge sample IDs to the ENSG IDs not in gencode v39
filtered_ensg <- filtered_ensg %>% 
  dplyr::left_join(fusion_names, by="Gene_symbol") %>% 
  dplyr::select(-FusionName) %>% 
  dplyr::distinct()

# merge OpenPedCan cohorts to the ENSG IDs not in gencode v39
filtered_ensg <- filtered_ensg %>% 
  dplyr::left_join(sample_ids, by = "Kids_First_Biospecimen_ID")

# write to file
readr::write_tsv(filtered_ensg,
  file.path(results_dir, 
            "putative-oncogene-fusion-freq_dropped_ensg.tsv.gz"))

# Display removed ENSEMBL IDs 
filtered_ensg %>% dplyr::select(Ensembl_ID, cohort) %>% 
  dplyr::distinct() %>%
  tidyr::drop_na() %>% 
  table()

# Display samples in removed ENSEMBL IDs that are in PMTL list
filtered_ensg %>% dplyr::filter(Ensembl_ID %in% pmtl_ids) %>% 
  select(Ensembl_ID, Kids_First_Biospecimen_ID) %>% 
  dplyr::distinct()

# Remove frequencies data frame
rm(fusion, filtered_ensg)
```

#### Fused gene frequencies
```{r}
# Remove Ensembl gene identifiers
fused_gene <- filter_freq_table(
  file.path(fusion_results_dir, 
            "putative-oncogene-fused-gene-freq.tsv.gz"), ensg_ids, ot_ensg_ids) 

# Ensembl gene identifiers removed
filtered_ensg <- fused_gene %>% 
  dplyr::filter(!targetFromSourceId %in% setdiff(ensg_ids, ot_ensg_ids)) %>% 
  dplyr::select(Gene_symbol, targetFromSourceId) %>%
  dplyr::rename(Ensembl_ID = targetFromSourceId) %>% 
  dplyr::distinct()

# merge sample IDs to the ENSG IDs not in gencode v39
filtered_ensg <- filtered_ensg %>% 
  dplyr::left_join(fusion_names, by = "Gene_symbol") %>% 
  dplyr::distinct()

# merge OpenPedCan cohorts to the ENSG IDs not in gencode v39
filtered_ensg <- filtered_ensg %>% 
  dplyr::left_join(sample_ids, by = "Kids_First_Biospecimen_ID")

# write to file
readr::write_tsv(filtered_ensg,
  file.path(results_dir, 
            "putative-oncogene-fused-gene-freq_dropped_ensg.tsv.gz"))

# Display removed ENSEMBL IDs 
filtered_ensg %>% dplyr::select(Ensembl_ID, cohort) %>% 
  dplyr::distinct() %>%
  tidyr::drop_na() %>% 
  table()

# Display samples in removed ENSEMBL IDs that are in PMTL list
filtered_ensg %>% dplyr::filter(Ensembl_ID %in% pmtl_ids) %>% 
  select(Ensembl_ID, Kids_First_Biospecimen_ID) %>% 
  dplyr::distinct()

# Remove frequencies data frame
rm(fused_gene, filtered_ensg, fusion_names)
```

## Session Info
```{r}
sessionInfo()
```